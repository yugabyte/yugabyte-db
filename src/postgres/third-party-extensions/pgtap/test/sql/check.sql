\unset ECHO
\i test/setup.sql

SELECT plan(48);

-- This will be rolled back. :-)
SET client_min_messages = warning;
CREATE TABLE public.sometab(
    id    INT NOT NULL PRIMARY KEY,
    name  TEXT DEFAULT '' CHECK ( name IN ('foo', 'bar', 'baz') ),
    numb  NUMERIC(10, 2),
    myint NUMERIC(8),
    CHECK (numb > 1.0 AND myint < 10)
);
RESET client_min_messages;

/****************************************************************************/
-- Test has_check().

SELECT * FROM check_test(
    has_check( 'public', 'sometab', 'public.sometab should have a check constraint' ),
    true,
    'has_check( schema, table, desc )',
    'public.sometab should have a check constraint',
    ''
);

SELECT * FROM check_test(
    has_check( 'sometab', 'sometab should have a check constraint' ),
    true,
    'has_check( table, desc )',
    'sometab should have a check constraint',
    ''
);

SELECT * FROM check_test(
    has_check( 'sometab' ),
    true,
    'has_check( table )',
    'Table sometab should have a check constraint',
    ''
);

SELECT * FROM check_test(
    has_check( 'pg_catalog', 'pg_class', 'pg_catalog.pg_class should have a check constraint' ),
    false,
    'has_check( schema, table, descr ) fail',
    'pg_catalog.pg_class should have a check constraint',
    ''
);

SELECT * FROM check_test(
    has_check( 'pg_class', 'pg_class should have a check constraint' ),
    false,
    'has_check( table, desc ) fail',
    'pg_class should have a check constraint',
    ''
);

/****************************************************************************/
-- Test col_has_check().

SELECT * FROM check_test(
    col_has_check( 'public', 'sometab', 'name', 'public.sometab.name should have a check' ),
    true,
    'col_has_check( sch, tab, col, desc )',
    'public.sometab.name should have a check',
    ''
);

SELECT * FROM check_test(
    col_has_check( 'public', 'sometab', ARRAY['numb', 'myint'], 'public.sometab.numb+myint should have a check' ),
    true,
    'col_has_check( sch, tab, cols, desc )',
    'public.sometab.numb+myint should have a check',
    ''
);

SELECT * FROM check_test(
    col_has_check( 'sometab', 'name', 'sometab.name should have a check' ),
    true,
    'col_has_check( tab, col, desc )',
    'sometab.name should have a check',
    ''
);

SELECT * FROM check_test(
    col_has_check( 'sometab', ARRAY['numb', 'myint'], 'sometab.numb+myint should have a check' ),
    true,
    'col_has_check( tab, cols, desc )',
    'sometab.numb+myint should have a check',
    ''
);

SELECT * FROM check_test(
    col_has_check( 'sometab', 'name' ),
    true,
    'col_has_check( table, column )',
    'Column sometab(name) should have a check constraint',
    ''
);

SELECT * FROM check_test(
    col_has_check( 'sometab', ARRAY['numb', 'myint'] ),
    true,
    'col_has_check( table, columns )',
    'Columns sometab(numb, myint) should have a check constraint',
    ''
);

SELECT * FROM check_test(
    col_has_check( 'public', 'sometab', 'id', 'public.sometab.id should have a check' ),
    false,
    'col_has_check( sch, tab, col, desc ) fail',
    'public.sometab.id should have a check',
    '        have: {name}
              {numb,myint}
        want: {id}'
);

SELECT * FROM check_test(
    col_has_check( 'sometab', 'id', 'sometab.id should have a check' ),
    false,
    'col_has_check( tab, col, desc ) fail',
    'sometab.id should have a check',
    '        have: {name}
              {numb,myint}
        want: {id}'
);

/****************************************************************************/
-- Test col_has_check() with an array of columns.

SET LOCAL client_min_messages = warning;
CREATE TABLE public.argh (
    id   INT  NOT NULL,
    name TEXT NOT NULL,
    CHECK ( id IN (1, 2) AND name IN ('foo', 'bar'))
);
RESET client_min_messages;

SELECT * FROM check_test(
    col_has_check( 'public', 'argh', ARRAY['id', 'name'], 'id + name should have a check' ),
    true,
    'col_has_check( sch, tab, col[], desc )',
    'id + name should have a check',
    ''
);

SELECT * FROM check_test(
    col_has_check( 'argh', ARRAY['id', 'name'], 'id + name should have a check' ),
    true,
    'col_has_check( tab, col[], desc )',
    'id + name should have a check',
    ''
);

SELECT * FROM check_test(
    col_has_check( 'argh', ARRAY['id', 'name'] ),
    true,
    'col_has_check( tab, col[] )',
    'Columns argh(id, name) should have a check constraint',
    ''
);

/****************************************************************************/
-- Finish the tests and clean up.
SELECT * FROM finish();
ROLLBACK;
