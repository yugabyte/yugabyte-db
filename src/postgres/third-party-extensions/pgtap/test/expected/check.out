\unset ECHO
1..48
ok 1 - has_check( schema, table, desc ) should pass
ok 2 - has_check( schema, table, desc ) should have the proper description
ok 3 - has_check( schema, table, desc ) should have the proper diagnostics
ok 4 - has_check( table, desc ) should pass
ok 5 - has_check( table, desc ) should have the proper description
ok 6 - has_check( table, desc ) should have the proper diagnostics
ok 7 - has_check( table ) should pass
ok 8 - has_check( table ) should have the proper description
ok 9 - has_check( table ) should have the proper diagnostics
ok 10 - has_check( schema, table, descr ) fail should fail
ok 11 - has_check( schema, table, descr ) fail should have the proper description
ok 12 - has_check( schema, table, descr ) fail should have the proper diagnostics
ok 13 - has_check( table, desc ) fail should fail
ok 14 - has_check( table, desc ) fail should have the proper description
ok 15 - has_check( table, desc ) fail should have the proper diagnostics
ok 16 - col_has_check( sch, tab, col, desc ) should pass
ok 17 - col_has_check( sch, tab, col, desc ) should have the proper description
ok 18 - col_has_check( sch, tab, col, desc ) should have the proper diagnostics
ok 19 - col_has_check( sch, tab, cols, desc ) should pass
ok 20 - col_has_check( sch, tab, cols, desc ) should have the proper description
ok 21 - col_has_check( sch, tab, cols, desc ) should have the proper diagnostics
ok 22 - col_has_check( tab, col, desc ) should pass
ok 23 - col_has_check( tab, col, desc ) should have the proper description
ok 24 - col_has_check( tab, col, desc ) should have the proper diagnostics
ok 25 - col_has_check( tab, cols, desc ) should pass
ok 26 - col_has_check( tab, cols, desc ) should have the proper description
ok 27 - col_has_check( tab, cols, desc ) should have the proper diagnostics
ok 28 - col_has_check( table, column ) should pass
ok 29 - col_has_check( table, column ) should have the proper description
ok 30 - col_has_check( table, column ) should have the proper diagnostics
ok 31 - col_has_check( table, columns ) should pass
ok 32 - col_has_check( table, columns ) should have the proper description
ok 33 - col_has_check( table, columns ) should have the proper diagnostics
ok 34 - col_has_check( sch, tab, col, desc ) fail should fail
ok 35 - col_has_check( sch, tab, col, desc ) fail should have the proper description
ok 36 - col_has_check( sch, tab, col, desc ) fail should have the proper diagnostics
ok 37 - col_has_check( tab, col, desc ) fail should fail
ok 38 - col_has_check( tab, col, desc ) fail should have the proper description
ok 39 - col_has_check( tab, col, desc ) fail should have the proper diagnostics
ok 40 - col_has_check( sch, tab, col[], desc ) should pass
ok 41 - col_has_check( sch, tab, col[], desc ) should have the proper description
ok 42 - col_has_check( sch, tab, col[], desc ) should have the proper diagnostics
ok 43 - col_has_check( tab, col[], desc ) should pass
ok 44 - col_has_check( tab, col[], desc ) should have the proper description
ok 45 - col_has_check( tab, col[], desc ) should have the proper diagnostics
ok 46 - col_has_check( tab, col[] ) should pass
ok 47 - col_has_check( tab, col[] ) should have the proper description
ok 48 - col_has_check( tab, col[] ) should have the proper diagnostics
